lib/concurrent-merkle-tree/src/merkle_roll.rs

```diff
@@ -210,109 +267,132 @@ impl<const MAX_DEPTH: usize, const MAX_BUFFER_SIZE: usize> MerkleRoll<MAX_DEPTH,
             }
             // If use_full_buffer is true, this loop will do 1 full pass of the change logs
-            if use_full_buffer && j == self.active_index {
+            if use_full_buffer && changelog_buffer_index == self.active_index {
                 break;
             }
         }
         log_compute!();
-        let valid_root = recompute(updated_leaf, proof, index) == self.get_change_log().root;
-        if updated_leaf != leaf || index > self.rightmost_proof.index {
+        updated_leaf == leaf
+    }
+
+    /// Update root & record changelog
+    /// --------
+    /// Fast-forwards submitted proof to be valid for the root at `self.current_index`
+    ///
+    /// Enabling `use_full_buffer` will skip root matching and just fast forward the given proof
+    /// from the beginning of the buffer.
+    /// Note: `use_full_buffer` significantly reduces security
+    fn apply_and_record_proof(
+        &mut self,
+        current_root: Node,
+        leaf: Node,
+        new_leaf: Node,
+        proof: &mut [Node; MAX_DEPTH],
+        leaf_index: u32,
+        append_on_conflict: bool,
+        use_full_buffer: bool,
+    ) -> Result<Node, CMTError> {
+        solana_logging!("Active Index: {}", self.active_index);
+        solana_logging!("Rightmost Index: {}", self.rightmost_proof.index);
+        solana_logging!("Buffer Size: {}", self.buffer_size);
+        solana_logging!("Leaf Index: {}", leaf_index);
+
+        let mask: usize = MAX_BUFFER_SIZE - 1;
+        let changelog_buffer_index: u64;
+        if use_full_buffer {
+            changelog_buffer_index = self.active_index.wrapping_sub(self.buffer_size) & mask as u64
+        } else {
+            match self.find_root_in_changelog(current_root) {
+                Some(matching_changelog_index) => {
+                    changelog_buffer_index = matching_changelog_index;
+                }
+                None => return Err(CMTError::RootNotFound),
+            }
+        };
+
+        let valid_fast_forward = self.fast_forward_proof(
+            new_leaf,
+            proof,
+            leaf_index,
+            changelog_buffer_index,
+            use_full_buffer,
+        );
+
+        let valid_root = recompute(leaf, proof, leaf_index) == self.get_change_log().root;
+        if !valid_fast_forward || leaf_index > self.rightmost_proof.index {
             // If the supplied root was not found in the queue, the instruction should fail if the leaf index changes
-            if !use_full_buffer && valid_root && leaf == EMPTY && append_on_conflict {
+            // NOTE: previously we checked if the FF'd proof with the value of the overwritten leaf
+            //      could be hashed to match the current root
+            //      However, this was removed for simplicity because it did not add to the security model
+            //      of insert_or_append functionality.
+            if !use_full_buffer && leaf == EMPTY && append_on_conflict {
```

Comment:
Can we add `valid_root` back here? What about the refactor we initially talked about?

Reply:
Which state refactors? Also I thought we couldn't find a situation where valid root check makes sense for insert_or_append? 

Are you saying I should refactor insert or append to first set_leaf, then append on failure?

Reply:
We synced off line. The answer is yes